source/id_descriptor.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "id_descriptor.h"
+
+#include <cassert>
+#include <iostream>
+
+#include "opcode.h"
+#include "operand.h"
+
+namespace libspirv {
+
+namespace {
+
+uint32_t HashU32Array(const std::vector<uint32_t>& words) {
+  const uint32_t kMagic = 2654435761;
+  uint32_t val = 0;
+  for (uint32_t i = 0; i < words.size(); ++i) {
+    val += (words[i] + i + 123) * kMagic;
+  }
+  return val;
+}
+
+}  // namespace
+
+uint32_t IdDescriptorCollection::IssueNewDescriptor(
+    const spv_parsed_instruction_t& inst) {
+  if (!inst.result_id)
+    return 0;
+
+  words_.clear();
```

Comment:
Minor:  Why not clear at the end rather than the beginning?  Leaving it at the end will mean you have data you'll never use again, lying around.